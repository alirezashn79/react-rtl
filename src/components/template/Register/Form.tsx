import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { zodResolver } from "@hookform/resolvers/zod";
import { SubmitHandler, useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { z } from "zod";

const formSchema = z
  .object({
    email: z.string().trim().email().min(4),
    password: z.string().trim().min(4),
    confirmPassword: z.string().trim().min(4),
  })
  .superRefine(({ password, confirmPassword }, ctx) => {
    if (password !== confirmPassword) {
      ctx.addIssue({
        code: "custom",
        message: "The passwords did not match",
        path: ["confirmPassword"],
      });
    }
  });

type FormData = z.infer<typeof formSchema>;

export default function RegisterForm() {
  const form = useForm<FormData>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      email: "",
      password: "",
      confirmPassword: "",
    },
  });

  const navigate = useNavigate();

  const onSubmit: SubmitHandler<FormData> = (values) => {
    navigate("/products", {
      state: values,
    });
  };
  return (
    <div>
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-4">
          <FormField
            control={form.control}
            name="email"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Email</FormLabel>
                <FormControl>
                  <Input placeholder="Email" {...field} />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="password"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Password</FormLabel>
                <FormControl>
                  <Input type="password" placeholder="Password" {...field} />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="confirmPassword"
            render={({ field }) => (
              <FormItem>
                <FormLabel>confirm Password</FormLabel>
                <FormControl>
                  <Input
                    type="password"
                    placeholder="confirm Password"
                    {...field}
                  />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />

          <Button type="submit">Submit</Button>
        </form>
      </Form>
    </div>
  );
}
