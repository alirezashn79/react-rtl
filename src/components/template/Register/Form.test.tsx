import Register from "@/pages/register";
import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";

interface formElementsType {
  Email: HTMLInputElement;
  Password: HTMLInputElement;
  ConfirmPassword: HTMLInputElement;
  Button: HTMLButtonElement;
}

const changeEvent = (element: HTMLInputElement, value: string) => {
  return act(() => userEvent.type(element, value));
};

const clickEvent = (element: HTMLButtonElement) => {
  return userEvent.click(element);
};

const mockUseNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockUseNavigate,
  };
});

const validData = {
  email: "alireza@test.com",
  password: "1234",
  confirmPassword: "1234",
};

const invalidData = {
  email: "invalid-email",
  password: "123",
  confirmPassword: {
    matchWithIncorrectPassword: "123",
    notMatch: "7896",
  },
};

const errorMessages = {
  email: "Invalid email",
  password: "String must contain at least 4 character(s)",
  confirmPassword: "The passwords did not match",
};

describe("Register Form", () => {
  let formElements: formElementsType;

  beforeEach(() => {
    render(
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    );
  });

  beforeEach(() => {
    formElements = {
      Email: screen.getByRole("textbox", { name: "Email" }),
      Password: screen.getByLabelText("Password"),
      ConfirmPassword: screen.getByLabelText("Confirm Password"),
      Button: screen.getByRole("button", { name: "Submit" }),
    };
  });

  test("should inputs be initially empty", () => {
    expect(formElements.Email).toHaveValue("");
    expect(formElements.Password).toHaveValue("");
    expect(formElements.ConfirmPassword).toHaveValue("");
  });

  test("should be able to type into inputs", async () => {
    await changeEvent(formElements.Email, "alireza@test.com");
    await changeEvent(formElements.Password, "12345");
    await changeEvent(formElements.ConfirmPassword, "12345");
  });

  test("should button be disabled when inputs are empty", () => {
    expect(formElements.Button).toHaveAttribute("type", "submit");
    expect(formElements.Button).toBeDisabled();
  });

  test("should button be enabled when all inputs are filled", async () => {
    await changeEvent(formElements.Email, "alireza@test.com");
    await changeEvent(formElements.Password, "12345");
    await changeEvent(formElements.ConfirmPassword, "12345");

    expect(formElements.Button).toBeEnabled();
  });

  describe("Handle Errors", () => {
    beforeEach(() => {
      expect(screen.queryByText(errorMessages.email)).not.toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.password)
      ).not.toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.confirmPassword)
      ).not.toBeInTheDocument();
    });

    test("should show email error message on invalid email", async () => {
      await changeEvent(formElements.Email, invalidData.email);
      await changeEvent(formElements.Password, validData.password);
      await changeEvent(
        formElements.ConfirmPassword,
        validData.confirmPassword
      );

      await clickEvent(formElements.Button);

      expect(screen.getByText(errorMessages.email)).toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.password)
      ).not.toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.confirmPassword)
      ).not.toBeInTheDocument();
    });

    test("should show password error message on invalid password", async () => {
      await changeEvent(formElements.Email, validData.email);
      await changeEvent(formElements.Password, invalidData.password);
      await changeEvent(
        formElements.ConfirmPassword,
        invalidData.confirmPassword.matchWithIncorrectPassword
      );

      await clickEvent(formElements.Button);

      expect(screen.queryByText(errorMessages.email)).not.toBeInTheDocument();
      expect(
        screen.getAllByText(errorMessages.password)[0]
      ).toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.confirmPassword)
      ).not.toBeInTheDocument();
    });

    test("should show confirm password error message on invalid confirm password", async () => {
      await changeEvent(formElements.Email, validData.email);
      await changeEvent(formElements.Password, validData.password);
      await changeEvent(
        formElements.ConfirmPassword,
        invalidData.confirmPassword.matchWithIncorrectPassword
      );

      await clickEvent(formElements.Button);

      expect(screen.queryByText(errorMessages.email)).not.toBeInTheDocument();
      expect(
        screen.getAllByText(errorMessages.password)[0]
      ).toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.confirmPassword)
      ).not.toBeInTheDocument();
    });

    test("should show confirm password error message on password not match", async () => {
      await changeEvent(formElements.Email, validData.email);
      await changeEvent(formElements.Password, validData.password);
      await changeEvent(
        formElements.ConfirmPassword,
        invalidData.confirmPassword.notMatch
      );

      await clickEvent(formElements.Button);

      expect(screen.queryByText(errorMessages.email)).not.toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.password)
      ).not.toBeInTheDocument();
      expect(
        screen.getByText(errorMessages.confirmPassword)
      ).toBeInTheDocument();
    });
  });

  describe("Handle Navigation", () => {
    test("should call the navigation after clicking on submit button", async () => {
      await changeEvent(formElements.Email, validData.email);
      await changeEvent(formElements.Password, validData.password);
      await changeEvent(
        formElements.ConfirmPassword,
        validData.confirmPassword
      );

      await clickEvent(formElements.Button);

      expect(screen.queryByText(errorMessages.email)).not.toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.password)
      ).not.toBeInTheDocument();
      expect(
        screen.queryByText(errorMessages.confirmPassword)
      ).not.toBeInTheDocument();

      expect(mockUseNavigate).toHaveBeenCalledWith("/products", {
        state: validData,
      });
    });
  });
});
