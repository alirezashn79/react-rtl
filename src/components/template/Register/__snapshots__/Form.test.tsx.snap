// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Register Form > Snapshot > should match with snapshop 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-screen w-screen flex items-center justify-center"
      >
        <div
          class="rounded-xl border bg-card text-card-foreground shadow min-w-96"
        >
          <div
            class="flex flex-col space-y-1.5 p-6"
          >
            <div
              class="font-semibold tracking-tight text-2xl text-center"
            >
              Register
            </div>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div>
              <form
                class="space-y-4"
                role="form"
              >
                <div
                  class="space-y-2"
                >
                  <label
                    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                    for=":rr:-form-item"
                  >
                    Email
                  </label>
                  <input
                    aria-describedby=":rr:-form-item-description"
                    aria-invalid="false"
                    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                    id=":rr:-form-item"
                    name="email"
                    placeholder="Email"
                    value=""
                  />
                </div>
                <div
                  class="space-y-2"
                >
                  <label
                    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                    for=":rs:-form-item"
                  >
                    Password
                  </label>
                  <input
                    aria-describedby=":rs:-form-item-description"
                    aria-invalid="false"
                    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                    id=":rs:-form-item"
                    name="password"
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                </div>
                <div
                  class="space-y-2"
                >
                  <label
                    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                    for=":rt:-form-item"
                  >
                    Confirm Password
                  </label>
                  <input
                    aria-describedby=":rt:-form-item-description"
                    aria-invalid="false"
                    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                    id=":rt:-form-item"
                    name="confirmPassword"
                    placeholder="Confirm Password"
                    type="password"
                    value=""
                  />
                </div>
                <button
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
                  disabled=""
                  type="submit"
                >
                  Submit
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="h-screen w-screen flex items-center justify-center"
      >
        <div
          class="rounded-xl border bg-card text-card-foreground shadow min-w-96"
        >
          <div
            class="flex flex-col space-y-1.5 p-6"
          >
            <div
              class="font-semibold tracking-tight text-2xl text-center"
            >
              Register
            </div>
          </div>
          <div
            class="p-6 pt-0"
          >
            <div>
              <form
                class="space-y-4"
                role="form"
              >
                <div
                  class="space-y-2"
                >
                  <label
                    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                    for=":ru:-form-item"
                  >
                    Email
                  </label>
                  <input
                    aria-describedby=":ru:-form-item-description"
                    aria-invalid="false"
                    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                    id=":ru:-form-item"
                    name="email"
                    placeholder="Email"
                    value=""
                  />
                </div>
                <div
                  class="space-y-2"
                >
                  <label
                    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                    for=":rv:-form-item"
                  >
                    Password
                  </label>
                  <input
                    aria-describedby=":rv:-form-item-description"
                    aria-invalid="false"
                    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                    id=":rv:-form-item"
                    name="password"
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                </div>
                <div
                  class="space-y-2"
                >
                  <label
                    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                    for=":r10:-form-item"
                  >
                    Confirm Password
                  </label>
                  <input
                    aria-describedby=":r10:-form-item-description"
                    aria-invalid="false"
                    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                    id=":r10:-form-item"
                    name="confirmPassword"
                    placeholder="Confirm Password"
                    type="password"
                    value=""
                  />
                </div>
                <button
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
                  disabled=""
                  type="submit"
                >
                  Submit
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-screen w-screen flex items-center justify-center"
    >
      <div
        class="rounded-xl border bg-card text-card-foreground shadow min-w-96"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <div
            class="font-semibold tracking-tight text-2xl text-center"
          >
            Register
          </div>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div>
            <form
              class="space-y-4"
              role="form"
            >
              <div
                class="space-y-2"
              >
                <label
                  class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                  for=":ru:-form-item"
                >
                  Email
                </label>
                <input
                  aria-describedby=":ru:-form-item-description"
                  aria-invalid="false"
                  class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                  id=":ru:-form-item"
                  name="email"
                  placeholder="Email"
                  value=""
                />
              </div>
              <div
                class="space-y-2"
              >
                <label
                  class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                  for=":rv:-form-item"
                >
                  Password
                </label>
                <input
                  aria-describedby=":rv:-form-item-description"
                  aria-invalid="false"
                  class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                  id=":rv:-form-item"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="space-y-2"
              >
                <label
                  class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                  for=":r10:-form-item"
                >
                  Confirm Password
                </label>
                <input
                  aria-describedby=":r10:-form-item-description"
                  aria-invalid="false"
                  class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
                  id=":r10:-form-item"
                  name="confirmPassword"
                  placeholder="Confirm Password"
                  type="password"
                  value=""
                />
              </div>
              <button
                class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
                disabled=""
                type="submit"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
